README.md
# Component Library Workshop

_a small starter for my 'Build a Component Library' Workshop at [DinosaurJS](https://dinosaurjs.org/)_

## Overview

This component lib is set up to use
[styled-components](https://github.com/styled-components/styled-components) and
[polished](https://github.com/styled-components/polished) 💅. This
starter is designed to help you create your own components and easily publish them to npm.

## Up & Running

To install dependencies with Yarn, run:

```sh
$ yarn
```

or to install with npm, run:

```sh
$ npm install
```

## File Structure

This component library borrows its structure from [BEM](http://getbem.com/introduction/) and is set
up to use Blocks, Elements, and Modifiers. Below is a description of each.

### Blocks

Blocks are the highest level of abstraction in the Blocks, Elements, Modifiers concept.
They are responsible for providing the context for Elements, handling UI logic, and rendering the
Elements within the Block. They are not connected to application state, nor do they handle any
business logic.

### Elements

Elements are the smallest, indivisible parts of UI. They are responsible for actually rendering the
UI. They do not handle application logic or UI logic, but they do handle their own modifiers which
modify the element’s style. Elements generally exist within the context of a Block (as their own
file in the Block’s directory) allowing the reuse of that set of Elements, but they are not
exclusively bound to blocks. An example of a stand-alone Element would be an A, Link, H3, or P.
These common elements live in `lib/elements`.

### Modifiers

This library utilizes
[`styled-components-modifiers`](https://github.com/Decisiv/styled-components-modifiers) to build
modifiers. Modifiers are small functions that allow us to alter the properties of an Element.
They primarily live in the Element's file and are solely responsible for modifying styles.
Some modifiers are common to multiple Elements. An example would be `fontWeights`.
These common modifiers live in `lib/modifiers`

### An Example Structure

```
├ lib/
├── blocks/
|   ├── Card
|   |   ├── Body.js     // <- Element
|   |   ├── Footer.js   // <- Element
|   |   ├── Header.js   // <- Element
|   |   └── index.js    // <- Block
|   └── index.js        // <- export for all Blocks
├── elements/
|   ├── A
|   |   ├── __tests__
|   |   |   ├── __snapshots__
|   |   |   |   └── index.js.snap   // <- Snapshot Test
|   |   |   └── index.js            // <- Test
|   |   └── index.js                // <- Element
|   ├── Link
|   |   └── index.js                // <- Element
|   ├── H3
|   |   └── index.js                // <- Element
|   ├── P
|   |   └── index.js                // <- Element
|   └── etc.
|   └── index.js                    // <- export for all Blocks
├── modifiers/
|   ├── fontWeights
|   └── etc.
└── index.js                        // <- main export for the library
```

## Local Development

### Module Development Workflow

Helpful information on development workflow in this library lives
[here](https://gist.github.com/alanbsmith/6c581e5042b8e5e558b0b4454192eb69).

### Linting

_**NOTE:** The linter will run against everything in the `lib` directory. I've added an initial
`.eslintrc` file for some basic configuration. Feel free to edit or replace it as needed. The
intent is to help give you a guide for syntax as you build your application. However, if the
linters are too distracting and / or confusing, feel free to ignore them._

### JavaScript Linting

_This assumes you have eslint and eslint-watch installed. If you don't, run the following:_

```
$ npm i -g eslint eslint-watch
```

or if you need permissions:

```
$ sudo npm i -g eslint eslint-watch
```

To run the linter once:

```
$ yarn lint:js
```

To run the watch task:

```
$ yarn lint:js:watch
```

#### Style Linting

I've also added a style linter for Sass / SCSS.

To run the style linter:

```
$ yarn lint:style
```

#### Linting JavaScript & Styles

To run both linters:

```
$ yarn lint
```

### Testing

An initial test suite has been setup with two tests (one passing and one intentionally failing).
We're using Jest Snapshots for our initial test setup, though Enzym and Expect are also available.
The basic test setup lives in `./__tests__`. The main configuration for Jest lives at the bottom
of `package.json`. I've also added a few handy scripts, which I've listed below. Jest also gives
us a test coverage tool for free, so I've added that too. The setup is at the bottom of
`package.json`. Everything is set to 90% coverage, but your welcome to update that to whatever
you'd like.

To run the tests once:

```
$ npm test
```

To run the watch script (for only relevant test files)

```
$ npm run test:watch
```

To run the watch script (for all test files)

```
$ npm run test:watchAll
```

To view the coverage report:

```
$ npm run test:coverage:report
```

### Review

If you'd like to run the linters and tests at once (this is a nice check before pushing to
Github or deploys), you can run:

```
$ npm run review
```

### Prettier

This library uses [Prettier](https://github.com/prettier/prettier) for code consistency. There's a pre-commit hook that will prettier and roll those changes into your commit.

### Docs

This lib uses [react-styleguidist](https://github.com/styleguidist/react-styleguidist) for documenting components. To to view the docs locally, run `yarn styleguide` and visit `http://localhost:6060`. To build a static version, run `yarn styleguide:build`. This static version will be created in `/docs`. You can tell GitHub to use this directory as the source for your docs. To do this, visit the Settings page and find the GitHub Pages section. Then select `master branch/docs folder` as your source and click "Save."

\_**NOTE:** There is a pre-push script in this library that will automatically update the docs when you push to GitHub.

### Build

_**NOTE:** When you run `build`, Babel will create a `build` directory. This is what your users
will interact with when they use your library. Nothing in `lib` gets shipped with your
published module._

Run once:

```
$ npm run build
```

Run the watch script:

```
$ npm run build:watch
```

_**NOTE:** the build script runs in the `prepublishOnly` script just before you publish to npm._

## Publishing

If you already have an account with npm, you can simply run:

```
$ npm login
$ npm publish
```

If you don't have an account with npm:

_**NOTE:** Your email address is public_

```
$ npm set init.author.name "Your Name"
$ npm set init.author.email "you@example.com"
$ npm set init.author.url "http://yourblog.com"
$ npm adduser
$ npm publish
```

## Contributing

I am thankful for any contributions made by the community. By contributing you agree to abide by
the Code of Conduct in the [Contributing Guidelines](https://github.com/alanbsmith/component-library-starter/blob/master/.github/CONTRIBUTING.md).

## License

[MIT](https://github.com/alanbsmith/component-library-starter/blob/master/LICENSE)


CHANGELOG.md
# Changelog

## 0.1.0

- Initial release 🎉
  - Add basic lib setup
  - Add Grid Blocks and related tests / snapshots


lib/blocks/Grid/__tests__/Column.js
import React from "react";
import Column from "../Column";
import "jest-styled-components";

import { renderWithTheme } from "__tests__/helpers";

describe("Grid.Column Element", () => {
  it("renders correctly", () => {
    const tree = renderWithTheme(<Column />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("applies the correct style with the `flex` prop", () => {
    const tree = renderWithTheme(<Column flex="1" />).toJSON();
    expect(tree).toMatchSnapshot();
    expect(tree).toHaveStyleRule("flex", "1");
  });
});


lib/blocks/Grid/__tests__/Container.js
import React from "react";
import Container from "../Container";
import "jest-styled-components";

import { renderWithTheme } from "__tests__/helpers";

describe("Grid.Container Element", () => {
  it("renders correctly", () => {
    const tree = renderWithTheme(<Container />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("applies the correct style with the `fluid` prop", () => {
    const tree = renderWithTheme(<Container fluid />).toJSON();
    expect(tree).toMatchSnapshot();
    expect(tree).toHaveStyleRule("padding", "0");
  });
});


config/theme.js
// common colors
const chrome000 = "#FFFFFF";
const chrome200 = "#A9BCC6";
const chrome300 = "#7B8493";
const chrome500 = "#24324C";
const chrome700 = "#101A2D";
const blue = "#4081F0";
const green = "#40F0B0";
const lightBlue = "#40D9F0";
const purple = "#5840F0";
const red = "#F05942";
const yellow = "#F0B040";

module.exports = {
  colors: {
    blue,
    green,
    lightBlue,
    purple,
    red,
    // ui colors
    ui: {
      background: chrome000,
      link: blue,
      linkHover: lightBlue,
      linkVisited: purple,
      text: chrome500,
      textlight: chrome300,
      // grayscale
      chrome000,
      chrome200,
      chrome300,
      chrome500,
      chrome700
    },
    // brand colors
    brand: {
      primary: chrome500,
      secondary: red,
      tertiary: "#FDE4A9"
    },
    // status colors
    status: {
      danger: red,
      info: blue,
      success: green,
      warning: yellow
    }
  },
  dimensions: {
    baseGrid: 8,
    borderRadius: "2px"
  },
  font: {
    primary: "Open Sans",
    primaryFallback: "Helvetica, Arial, sans-serif",
    weights: {
      light: 300,
      normal: 400,
      semiBold: 600,
      bold: 700
    },
    sizes: {
      extraSmall: 12,
      small: 14,
      medium: 16,
      large: 18,
      h1: 40,
      h2: 36,
      h3: 32,
      h4: 28,
      h5: 24,
      h6: 20
    }
  }
};


styleguide.config.js
const path = require("path");
const theme = require("./config/theme");

const fontSource =
  "https://fonts.googleapis.com/css?family=Open+Sans:300,400,600,700";

module.exports = {
  resolver: require("react-docgen").resolver.findAllComponentDefinitions,
  skipComponentsWithoutExample: true,
  styleguideComponents: {
    Wrapper: path.join(__dirname, "styleguide/ThemeWrapper")
  },
  styleguideDir: "docs",
  template: {
    head: {
      links: [
        {
          rel: "stylesheet",
          href: fontSource
        }
      ]
    }
  },
  ignore: ["**/utils/**.js"],
  sections: [
    {
      name: "Blocks",
      ignore: ["**/__tests__/**", "**/blocks/index.js"],
      components: "lib/blocks/**/index.js",
      description:
        "Blocks are the highest level of abstraction in the Blocks, Elements, Modifiers concept. They are responsible for providing the context for Elements, handling UI logic, and rendering the Elements within the Block. They are not connected to application state, nor do they handle any business logic."
    },
    {
      name: "Elements",
      ignore: ["**/__tests__/**", "**/elements/index.js"],
      components: "lib/elements/**/index.js",
      description:
        "Elements are the smallest, indivisible parts of UI. They are responsible for actually rendering the UI. They do not handle application logic or UI logic, but they do handle their own modifiers which modify the element’s style. Elements generally exist within the context of a Block (as their own file in the Block’s directory) allowing the reuse of that set of Elements, but they are not exclusively bound to blocks. An example of a stand-alone Element would be an A, Link, H3, or P. These common elements live in `lib/elements`."
    }
  ],
  theme: {
    color: {
      base: theme.colors.ui.text,
      light: theme.colors.ui.textLight,
      lightest: theme.colors.ui.chrome200,
      link: theme.colors.ui.link,
      linkHover: theme.colors.ui.linkHover,
      border: theme.colors.ui.chrome200,
      name: theme.colors.green,
      type: theme.colors.purple,
      error: theme.colors.red,
      baseBackground: theme.colors.ui.chrome000,
      codeBackground: theme.colors.ui.chrome100,
      sidebarBackground: theme.colors.ui.chrome100
    },
    fontFamily: {
      base: theme.font.primary
    },
    fontSize: {
      base: 12,
      text: 16,
      small: 14,
      h1: 40,
      h2: 36,
      h3: 32,
      h4: 28,
      h5: 24,
      h6: 20
    }
  }
};
